### ISSUE TITLE
Collect code coverage for test annotated with @coversNothing

### ISSUE THREAD

**Issue by Slamdunk** [sebastianbergmann/php-code-coverage#726]
Original issue: https://github.com/sebastianbergmann/phpunit/issues/3796

| Q                         | A
| --------------------------| ---------------
| php-code-coverage version | 8.0.0
| PHP version               | 7.3.14
| Driver                    | PCOV
| PCOV version (if used)  | 1.0.6
| Installation Method       | Composer
| Usage Method              | PHPUnit
| PHPUnit version (if used) | 9.0.0

Hi, https://github.com/sebastianbergmann/phpunit/pull/2560 introduced a performance improvement to skip code coverage collection when a test is marked with `@coversNothing`, but this generates a mixed report with `processUncoveredFilesFromWhitelist="true"`.

The real world example is when tests are executed in parallel, code coverage collected separately and merged back together in the end:

#### `src/MyClass.php`
```php
class MyClass
{
    public function foo()
    {
        return true;
    }
}
```
#### `tests/FirstTest.php`
```php
/**
 * @coversNothing
 */
class FirstTest extends PHPUnit\Framework\TestCase
{
    public function testOne()
    {
        static::assertTrue(true);
    }
}
```
#### `tests/SecondTest.php`
```php
/**
 * @covers MyClass
 */
class SecondTest extends PHPUnit\Framework\TestCase
{
    public function testOne()
    {
        static::assertTrue((new MyClass())->foo());
    }
}
```
#### `phpunit.xml`
```xml
<phpunit
    bootstrap="./vendor/autoload.php"
    backupGlobals="false"
    backupStaticAttributes="false"
    colors="true"
>
    <testsuite name="bug">
        <directory>./tests</directory>
    </testsuite>

    <filter>
        <whitelist processUncoveredFilesFromWhitelist="true">
            <directory suffix=".php">./src</directory>
        </whitelist>
    </filter>
</phpunit>
```

### Actual result with a single full run
![phpunit-full](https://user-images.githubusercontent.com/152236/63346937-5c9e6900-c356-11e9-8580-bacbb411229d.png)

### Actual result with the two tests executed on separate runs, merged together
![phpunit-parallel-merge](https://user-images.githubusercontent.com/152236/63346971-69bb5800-c356-11e9-9a7f-1a34514fc47a.png)

The reason should be that the lonely run of `FirstTest` never triggers `\SebastianBergmann\CodeCoverage\CodeCoverage::start`.

Side note: the reason behind https://github.com/sebastianbergmann/phpunit/pull/2560 can be considered superseded when PCOV is used, imho.

**Comment by sebastianbergmann** [comment]
I wish you had not simply copy/paste-ed the issue here from the PHPUnit repository. In the context of this repository, this issue needs to be framed around the fact that merging the data from two `CodeCoverage` objects does not work as expected.

**Comment by Slamdunk** [comment]
> I wish you had not simply copy/paste-ed the issue here from the PHPUnit repository.

I had no idea on how to improve the issue report, but I'd take few minutes to check if it's still an issue with the latest versions available; the Q/A table at the top differs.

**Comment by sebastianbergmann** [comment]
That would be very much appreciated, thanks.

**Comment by viezel** [comment]
I can second this issue. We are merging 9 reports together and this occurs.

**Comment by dekker-m** [comment]
I've seen the same behavior when we switched to pcov from xdebug a couple of days ago. We use Codeception to run the test cases, but in the end the coverage information is processed by php-code-coverage.

I checked different setups with different settings (xdebug/pcov and processUncoveredFilesFromWhitelist/addUncoveredFilesFromWhitelist). I still can't explain WHY this is happening, but as soon as addUncoveredFilesFromWhitelist is set to "false", the issue with pcov seems to be resolved. Could you guys check if setting addUncoveredFilesFromWhitelist to false resolve the issue on your side as well? Personally, I think adding the uncovered files from whitelists is redundant if processUncoveredFilesFromWhitelist is used as well. processUncoveredFilesFromWhitelist seems to do a better job anyway, because it uses the xdebug/pcov to detect executable lines.

FYI, the related issue in Codeception: https://github.com/Codeception/Codeception/issues/5768

**Comment by dekker-m** [comment]
*TL;DR:* Make sure `addUncoveredFromWhitelist` is set to `false` whenever you merge code coverage reports. Still want all uncovered files from you whitelist? Set `processUncoveredFilesFromWhitelist` to `true`.

I did some additional debugging. I printed the code coverage information to the console every time:

* Data for a specific file was merged using `merge()`
* Data for a specific file was appended using `append()`

The lines in the example of @Slamdunk marked red (6 & 8) are added through the `append()` method with test ID UNCOVERED_FILES_FROM_WHITELIST. A stack trace showed those lines were marked as "not executed" by the method `addUncoveredFilesFromWhitelist()`. This method is called every time a report is generated (intermediate or final). 

When the reports are merged, those lines marked as "not executed" are merged with the code coverage data from the test cases. I don't think the problem here is the `merge()` method itself, but the fact that those lines are added to one of the original reports in the first place by `addUncoveredFilesFromWhitelist()`.

There still is one thing which doesn't add up. `addUncoveredFilesFromWhitelist = true` should not add any files at all when `processUncoveredFilesFromWhitelist = true`, right? Well, this is not completely true. `processUncoveredFilesFromWhitelist` uses `include_once` to process the files using pcov or xdebug. If you have two instances of `CodeCoverage` in a single process and the second instance is initialized, `processUncoveredFilesFromWhitelist` does not work. So eventually `addUncoveredFilesFromWhitelist()` is called and you will end up with the "non executed" lines which are not executable after all.

The solution would be to set `addUncoveredFilesFromWhitelist` to `false` on all `CodeCoverage` instances used in generating the merged CodeCoverage report.

**Comment by sebastianbergmann** [comment]
@Slamdunk Is this still a problem for you?

**Comment by Slamdunk** [comment]
PHPUnit 9.4 with PHP-Code-Coverage 9.2 fixed the issue, hurray :rocket: 

I'd like to add a test for this case to avoid reintroducing this in the future, though, before closing the issue; I'd need few weeks for this as I'm busy right now.
